cmake-edit-core/src/main/java/ru/urururu/cmakeedit/core/checker/SimulationState.java

```diff
@@ -12,119 +11,149 @@
-    private void processUsage(String argument, Set<Node> suspiciousPoints) {
-        Set<CommandInvocationNode> commandInvocationNode = variables.get(argument);
+    void processUsage(String variable) {
+        Set<CommandInvocationNode> nodes = variables.get(variable);
 
-        if (commandInvocationNode == null) {
-            return;
+        if (nodes == null) {
+            processUnknownUsage(variable);
+        } else {
+            suspiciousPoints.removeAll(nodes);
         }
+    }
+
+    void putValue(ArgumentNode argumentNode, CommandInvocationNode command, boolean parentScope) {
+        putValue(argumentNode, command, Collections.emptyMap(), parentScope);
+    }
+
+    void putValue(ArgumentNode argumentNode, CommandInvocationNode command, Map<String, Node> substitutions, boolean parentScope) {
+        // todo check that expressions from argumentNode are used
+
+        // todo inline argumentNode via processUsage()
+        String variable = getValue(argumentNode, substitutions);
+
+        putValue(command, variable, parentScope);
+    }
+
+    protected void putValue(CommandInvocationNode command, String variable, boolean parentScope) {
+        variables.put(variable, Collections.singleton(command));
+    }
+
+    static String getArgument(Node argumentNode) {
+        StringBuilder sb = new StringBuilder();
+
+        argumentNode.visitAll(new NodeVisitorAdapter() {
+            @Override
+            public void accept(ExpressionNode node) {
+                if (node.getKey() != null) {
+                    sb.append(node.getKey());
+                }
+                //throw new IllegalStateException("not implemented");
+                node.getNested().stream().forEach(n -> n.visitAll(this));
+            }
+
+            @Override
+            public void accept(ConstantNode node) {
+                sb.append(node.getValue());
+            }
+        });
 
-        suspiciousPoints.removeAll(commandInvocationNode);
+        return sb.toString();
+    }
+
+    void simulate(Node node) {
+        node.visitAll(new NodeVisitorAdapter() {
+            @Override
+            public void accept(CommandInvocationNode node) {
+                processUsages(node);
+
+                if (node.getCommandName().equals("unset")) {
+                    List<Node> arguments = node.getArguments();
+                    if (arguments.size() > 0) {
+                        String variable = getArgument(arguments.get(0));
+                        // todo if variable is a reference through other variables we should inline it.
+
+                        variables.remove(variable);
+                    }
+                }
+            }
+
+            private void processUsages(CommandInvocationNode node) {
+                List<Node> arguments = node.getArguments();
+
+                for (Node argument : arguments) {
+                    String arg = getValue(argument);
+                    processUsage(arg); // todo shouldnt do for all!
```

Comment:
![INFO](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-info.png) Complete the task associated to this TODO comment. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://sonarqube.com/coding_rules#rule_key=squid%3AS1135)